Use Hyperlink widget for links in message info popup.

theViz343 opened the pull request:
<!-- See README for documentation, or ask in #zulip-terminal if unclear -->
### What does this PR do, and why?
This PR introduces the [Hyperlink](https://urwidgets.readthedocs.io/en/stable/api.html#urwidgets.Hyperlink) widget for message links in message information popups. This solves the issue of long urls being force-wrapped which results in them not being clickable from ZT.

### Outstanding aspect(s)    <!-- DELETE SECTION IF EMPTY -->
<!-- In what ways is this not fully implemented/functioning? Compared to a discussion/issue? -->
<!-- Do you not understand something? Are you unsure about a certain approach? Want feedback? -->
- The focus on the currently selected link is a bit off. It doesn't cover the link.

### External discussion & connections
<!-- [x] all that apply, specifying topic and adding numbers after # for issues/PRs -->
- [x] Discussed in **#zulip-terminal** in [`Hyperlink support using OSC-8 escape sequences`](https://chat.zulip.org/#narrow/stream/206-zulip-terminal/topic/Hyperlink.20support.20using.20OSC-8.20escape.20sequences)
- [ ] Fully fixes #
- [x] Partially fixes issue #1368 
- [ ] Builds upon previous unmerged work in PR #
- [ ] Is a follow-up to work in PR #
- [ ] Requires merge of PR #
- [ ] Merge will enable work on #

### How did you test this?
<!-- [x] all that apply -->
- [x] Manually - Behavioral changes
- [x] Manually - Visual changes
- [ ] Adapting existing automated tests
- [ ] Adding automated tests for new behavior (or missing tests)
- [ ] Existing automated tests should already cover this (*only a refactor of tested code*)

### Self-review checklist for each commit
- [x] It is a [minimal coherent idea](https://github.com/zulip/zulip-terminal#structuring-commits---speeding-up-reviews-merging--development)
- [x] It has a commit summary following the [documented style](https://github.com/zulip/zulip-terminal#structuring-commits---speeding-up-reviews-merging--development) (title & body)
- [x] It has a commit summary describing the  motivation and reasoning for the change
- [ ] It individually passes linting and tests
- [ ] It contains test additions for any new behavior
- [ ] It flows clearly from a previous branch commit, and/or prepares for the next commit

### Visual changes    <!-- DELETE SECTION IF NO VISUAL CHANGE -->
<!-- Zulip tips at https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html -->
<!-- For video, try asciinema; after uploading, embed using
[![yourtitle](https://asciinema.org/a/<id>.png)](https://asciinema.org/a/<id>)
-->
<!-- NOTE: Attached videos/images will be clearer from smaller terminal windows -->
![image](https://github.com/zulip/zulip-terminal/assets/44305101/4635aa0a-35f7-4989-a2ba-f0f181687a51)

theViz343 committed:
buttons/views: Use Hyperlink widget for links in message info popup.

This commit introduces the use of the Hyperlink widget from the
Urwidgets library for displaying links in the message information popup.

Subhasish-Behera commented:
Hey @theViz343 ,I am getting the error
```
 File "/home/subhasishbehera/zulip-terminal/zulipterminal/ui_tools/buttons.py", line 11, in <module>
    import urwidgets
ModuleNotFoundError: No module named 'urwidgets'

```

neiljp commented:
@Subhasish-Behera If you're using `make` then run that and it should check you have the required dependencies, or reinstall the editable zt package over the top (if not using make).

@theViz343 It would be useful to have the dependency change visible as a separate small commit, as we do when we change other `requirements`.

theViz343 committed:
requirements: setup: Add urwidgets library for hyperlink support.

This commit adds the urwidgets library for including hyperlink support
in ZT using the OSC-8 escape sequence.

neiljp commented on zulipterminal/ui_tools/buttons.py:
```diff
@@ -442,7 +443,12 @@ def update_widget(self, caption: str, display_attr: Optional[str] = None) -> Non
         """
         # Set cursor position next to len(caption) to avoid the cursor.
-        icon = urwid.SelectableIcon(caption, cursor_position=len(caption) + 1)
+        icon = urwid.Pile(
+            widget_list=[
+                urwid.SelectableIcon(caption, cursor_position=len(caption) + 1),
+                urwidgets.Hyperlink(uri=self.link),
+            ]
+        )
```
Having the rendering of the link defined in the link button like this would be cleaner, and it may be reasonable to refactor the layout into the button as a prep refactor for this.

However, right now this overall change affects the rendering quite a lot: links are not styled in the same way as the background of the button, or when selected. That makes it difficult to visually identify that the link and number/title are related to the link when many links are present.

I'm not sure if that's a limitation of the Hyperlink?

AnonymouX47 replied:
Hello!

> However, right now this overall change affects the rendering quite a lot: links are not styled in the same way as the background of the button, or when selected. That makes it difficult to visually identify that the link and number/title are related to the link when many links are present.
>
> I'm not sure if that's a limitation of the Hyperlink?

Yes, it was.

I just released [v0.2.0](https://github.com/AnonymouX47/urwidgets/releases/tag/v0.2.0) which allows the use of `Attrmap`s (and other means) over `Hyperlink` widgets (and the rendered canvases). See https://github.com/AnonymouX47/urwidgets/pull/3.

AnonymouX47 replied:
Replace with:

```py
        hyperlink = urwidgets.Hyperlink(uri=self.link)
        icon = urwid.Pile(
            widget_list=[
                urwid.SelectableIcon(caption, cursor_position=len(caption) + 1),
                # using `hyperlink.pack()[0]` instead of `len(self.link)` in case of wide characters
                urwidgets.TextEmbed((hyperlink.pack()[0], hyperlink)),
            ]
        )
```
to wrap the URL across lines instead of clipping with ellipsis.

The `OSC 8` feature will still work perfectly on any terminal emulator with a spec-compliant implementation and behaviour will remain the same as before this PR on terminal emulators that don't support the feature.
